test/unit/org/apache/cassandra/net/CallbackPropertyTest.java

```diff
@@ -0,0 +1,750 @@
+
+    private static final String DC1 = "datacenter1";
+    private static final String DC2 = "datacenter2";
+
+    /** By default, we'll test the CL's listed here unless the runner otherwise constrains things */
+    private EnumSet<ConsistencyLevel> testedCL = EnumSet.of(ConsistencyLevel.ALL, ConsistencyLevel.QUORUM, ConsistencyLevel.ONE);
+
+    public static DecoratedKey dk;
+
+    private SinglePartitionReadCommand command;
+
+    private ReplicaPlan.Shared readReplicaPlan;
+    private ReplicaPlan.ForWrite writeReplicaPlan;
+    private ReplicaPlan.ForPaxosWrite paxosReplicaPlan;
+    private ReplicaPlan.ForWrite eachQuorumPlan;
+    private ReplicaPlan.ForWrite localQuorumPlan;
+
+    private DataResolver resolver;
+    private Dispatcher.RequestTime requestTime;
+
+    private int dc1NodesDown;
+    private int dc2NodesDown = 0;
+    private int rf;
+    private ConsistencyLevel cl;
+    private int quorum;
+
+    /** We calculate whether we expect a test to succeed in the runner to then compare against from inside the test case's context */
+    private boolean caseExpectsException;
+
+    private Runnable testRunnable;
+
+    @BeforeClass
+    public static void setupClass()
+    {
+        DatabaseDescriptor.daemonInitialization();
+        DatabaseDescriptor.setPartitionerUnsafe(Murmur3Partitioner.instance);
+
+        // No need to wait so long
+        DatabaseDescriptor.setReadRpcTimeout(100);
+        DatabaseDescriptor.setWriteRpcTimeout(100);
+
+        ServerTestUtils.prepareServerNoRegister();
+
+        ClusterMetadataTestHelper.register(EP1, DC1, "R1");
+        ClusterMetadataTestHelper.register(EP2, DC1, "R1");
+        ClusterMetadataTestHelper.register(EP3, DC1, "R1");
+        ClusterMetadataTestHelper.register(EP4, DC1, "R1");
+        ClusterMetadataTestHelper.register(EP5, DC1, "R1");
+        ClusterMetadataTestHelper.register(EP6, DC2, "R2");
+        ClusterMetadataTestHelper.register(EP7, DC2, "R2");
+        ClusterMetadataTestHelper.register(EP8, DC2, "R2");
+        ClusterMetadataTestHelper.register(EP9, DC2, "R2");
+        ClusterMetadataTestHelper.register(EP10, DC2, "R2");
+        // We need to add endpoint's based on 0-4 == datacenter1, 5-9 == datacenter2.
+        for (int i = 0; i < 5; i++)
+        {
+            InetAddressAndPort peer1 = UNIQUE_EP.get(i);
+            InetAddressAndPort peer2 = UNIQUE_EP.get(i + 5);
+
+            Token t1 = token(i);
+            Token t2 = token(i + 5);
+
+            ClusterMetadataTestHelper.join(peer1, t1);
+            ClusterMetadataTestHelper.join(peer2, t2);
+        }
+    }
+
+    private void runTestMatrix(CallbackType type, Runnable r)
+    {
+        int minRF = 1;
+        testRunnable = r;
+
+        // Based on the type of callback we're testing we'll constrain various properties
+        switch (type)
+        {
+            case PAXOS:
+                minRF = 2;
+                testedCL = EnumSet.of(ConsistencyLevel.SERIAL);
```

Comment:
Won't overwriting the class level testedCLs mean the tested CLs are dependent on the order `runTestMatrix` is called if it isn't reset for each `type`?

Reply:
... I believe you're right. Think we need to set that back in the NORMAL and default case to the default value of:
```java
    private EnumSet<ConsistencyLevel> defaultCLTests = EnumSet.of(ConsistencyLevel.ALL, ConsistencyLevel.QUORUM, ConsistencyLevel.ONE);
    private EnumSet<ConsistencyLevel> testedCL = EnumSet.of(ConsistencyLevel.ALL, ConsistencyLevel.QUORUM, ConsistencyLevel.ONE);
```

I originally wrote these tests as 1 property test per file but then realized I could easily co-locate them all in the same file, but that certainly shifted the implications in terms of durable static state now didn't it? 🤔